tomviz/RotateAlignWidget.cxx

```diff
@@ -0,0 +1,319 @@
+#include "vtkDataArray.h"
+#include "vtkImageData.h"
+#include "vtkImageProperty.h"
+#include "vtkImageSlice.h"
+#include "vtkImageSliceMapper.h"
+#include "vtkInteractorStyleRubberBand2D.h"
+#include "vtkLineSource.h"
+#include "vtkNew.h"
+#include "vtkPointData.h"
+#include "vtkPolyDataMapper.h"
+#include "vtkProperty.h"
+#include "vtkRenderer.h"
+#include "vtkRenderWindow.h"
+#include "vtkScalarsToColors.h"
+#include "vtkSMSourceProxy.h"
+#include "vtkTransform.h"
+#include "vtkTrivialProducer.h"
+#include "vtkVector.h"
+
+#include "ui_RotateAlignWidget.h"
+
+#include <QGridLayout>
+#include <QHBoxLayout>
+#include <QVBoxLayout>
+#include <QLabel>
+#include <QPushButton>
+#include <QLineEdit>
+#include <QSpinBox>
+#include <QPointer>
+#include <QDoubleSpinBox>
+
+namespace
+{
+void setupRenderer(vtkRenderer *renderer, vtkImageSliceMapper *mapper)
+{
+  renderer->SetBackground(1.0, 1.0, 1.0);
+  vtkCamera *camera = renderer->GetActiveCamera();
+  renderer->SetViewport(0.0, 0.0,
+                        1.0, 1.0);
+
+  double *bounds = mapper->GetBounds();
+  vtkVector3d point;
+  point[0] = 0.5 * (bounds[0] + bounds[1]);
+  point[1] = 0.5 * (bounds[2] + bounds[3]);
+  point[2] = 0.5 * (bounds[4] + bounds[5]);
+  camera->SetFocalPoint(point.GetData());
+  point[2] += 500;
+  camera->SetPosition(point.GetData());
+  camera->SetViewUp(0.0, 1.0, 0.0);
+  camera->ParallelProjectionOn();
+  camera->SetParallelScale(0.5 * (bounds[1] - bounds[0] + 1));
+}
+}
+
+namespace tomviz
+{
+
+class RotateAlignWidget::RAWInternal
+{
+public:
+  Ui::RotateAlignWidget Ui;
+  vtkNew<vtkImageSlice> mainSlice;
+  vtkNew<vtkImageSlice> reconSlice[3];
+  vtkNew<vtkImageSliceMapper> mainSliceMapper;
+  vtkNew<vtkImageSliceMapper> reconSliceMapper[3];
+  vtkNew<vtkRenderer> mainRenderer;
+  vtkNew<vtkRenderer> reconRenderer[3];
+  vtkNew<vtkImageData> reconImage;
+  QPointer<DataSource> Source;
+  vtkNew<vtkLineSource> rotationAxis;
+  vtkNew<vtkActor> axisActor;
+
+  void setupCameras()
+  {
+    setupRenderer(this->mainRenderer.Get(), this->mainSliceMapper.Get());
+    setupRenderer(this->reconRenderer[0].Get(), this->reconSliceMapper[0].Get());
+    setupRenderer(this->reconRenderer[1].Get(), this->reconSliceMapper[1].Get());
+    setupRenderer(this->reconRenderer[2].Get(), this->reconSliceMapper[2].Get());
+  }
+
+  void setupRotationAxisLine()
+  {
+    vtkTrivialProducer *t =
+        vtkTrivialProducer::SafeDownCast(this->Source->producer()->GetClientSideObject());
+    if (!t)
+    {
+      return;
+    }
+    vtkImageData *imageData = vtkImageData::SafeDownCast(t->GetOutputDataObject(0));
+    if (imageData)
+    {
+      int extent[6];
+      imageData->GetExtent(extent);
+      double spacing[3];
+      imageData->GetSpacing(spacing);
+      int projections = extent[5] - extent[4] + 1;
+      this->Ui.projection->setMaximum(projections - 1);
+      this->Ui.rotationAxis->setRange(-projections * spacing[1] / 2.0, projections * spacing[1] / 2.0);
```

Comment:
The range of the rotationAxis should be  +/-(extent[3] - extent[2])/2 instead of +/-(extent[5] - extent[4])/2.
We are shifting the images in the y-direction.